src/clj/puppetlabs/services/request_handler/request_handler_core.clj

```diff
@@ -73,30 +101,25 @@
   ; the ring functions happen to slurp it up first.  This would happen for a
   ; 'application/x-www-form-urlencoded' form post where ring needs to slurp
   ; in the body of the request in order to parse out parameters from the form.
-
-  ; Arguably could use "ISO-8859-1" as the default encoding if one is not
-  ; specified on the request, but "UTF-8" is what ring would use as well.
-  (let [body-string (slurp (:body request)
-                           :encoding (or (:character-encoding request)
-                                         "UTF-8"))]
-        (->
-          request
-          ; Leave the slurped content under an alternate key so that it
-          ; is available to be proxied on to the JRubyPuppet request.
-          (assoc :body-string body-string)
-          ; Body content has been slurped already so wrap it in a new reader
-          ; so that a copy of it can be obtained by ring middleware functions,
-          ; if needed.
-          (assoc :body (StringReader. body-string))
-          ; Compojure request may have destructured parameters from subportions
-          ; of the URL into the params map by this point.  Clear this out
-          ; before invoking the ring middleware param functions so that keys
-          ; pulled from the query string or form body parameters don't
-          ; inadvertently conflict.
-          (assoc :params {})
-          ; Defer to ring middleware to pull out parameters from the query
-          ; string and/or form body.
-          ring-params/params-request)))
+  (let [body-for-jruby (body-for-jruby request)]
+    (->
+      request
+      ; Leave the slurped content under an alternate key so that it
+      ; is available to be proxied on to the JRubyPuppet request.
+      (assoc :body-for-jruby body-for-jruby)
+      ; Body content has been slurped already so wrap it in a new reader
+      ; so that a copy of it can be obtained by ring middleware functions,
+      ; if needed.
+      (assoc :body (if (string? body-for-jruby) (StringReader. body-for-jruby)))
```

Comment:
It would be better to avoid constructing the Java object, but I don't think the `string-reader` function that you mentioned will work.  The object is ultimately consumed by a ring middleware function, `ring.middleware.params/assoc-form-params`, which uses Clojure's `slurp`.  `slurp` tries to coerce the argument into a Java `InputStream` and fails with an `IllegalArgumentException`:

```
IllegalArgumentException Cannot open <#<StringReader clojure.tools.reader.reader_types.StringReader@53e7ec65>> as an InputStream.  clojure.java.io/fn--8616 (io.clj:173)
```

Would be nice if `slurp` knew how to consume a `clojure.tools.reader.reader_types.StringReader` properly, but looks like we're stuck here...